File: solution/go/enrichment/dns_tunnel.go
package main

import (
	"context"
	"log"

	enpb "github.com/bearlyrunning/FindingTheNeedle/solution/go/generated/enrichmentpb"
	spb "github.com/bearlyrunning/FindingTheNeedle/solution/go/generated/signalpb"
)

func (dte *DNSTunnelEnricher) getName() string {
	return dte.name
}

func (dte *DNSTunnelEnricher) getSignals() []*spb.Signal {
	return dte.signals
}

func (dte *DNSTunnelEnricher) setSignals(sigs []*spb.Signal) {
	dte.signals = sigs
}

// Skipping unit test as this involves setting up mock which is out of scope for this workshop.
func (dte *DNSTunnelEnricher) enrich(ctx context.Context, c enpb.EnrichmentClient) {
	for _, sig := range dte.signals {
		resp, err := c.IPToHost(ctx, &enpb.IP{Ip: sig.GetDnsTunnel().SourceIp})
		if err != nil {
			log.Printf("failed IPToHost conversion: %v", err)
		} else {
			sig.GetDnsTunnel().Hostname = resp.GetName()
		}
	}
}

File: solution/python/enrichment/dns_tunnel.py
import enricher
import logging
from generated import enrichment_pb2 as enpb
from generated import signal_pb2 as spb

class DNSTunnelEnricher(enricher.EnricherInterface):
    def __init__(self, name="", path=""):
        self.signals = []
        self.name = name
        self.signal_path = path

    # Get enricher name.
    def getName(self) -> str:
        return self.name
    
    # Get signals associated with a given enricher.
    def getSignals(self) -> list[spb.Signal]:
        return self.signals
    
    # Set the signals to be enriched.
    def setSignals(self, signals=[]) -> None:
        self.signals = signals

    # Enrich signals based on the provided client.
    def enrich(self, client) -> None:
        # For each signal, call our IPToHost enrichment.
        for sig in self.signals:
            try:
                resp = client.IPToHost(enpb.IP(
                    ip=sig.dns_tunnel.source_ip
                ))
                # Assign enriched hostname to protobuf field.
                sig.dns_tunnel.hostname = resp.name
            except Exception as e:
                logging.warn("failed IPToHost conversion: %s", str(e))